### DISCUSSION TITLE
YAML support for configuration file

### DISCUSSION THREAD

**Discussion by dfleury2** [CLIUtils/CLI11#724] [Ideas]
Hi,

I was wondering if there is plan to introduce YAML support for configuration file.
Will it be easy to do or not  (without adding external dependecy) ?

Regards,

**Comment by phlptp** [comment]
no plans, to generally support yaml.  Yaml files can get pretty complex so the full parser for them is rather complicated to handle the whitespace rules.  
A trivial yaml file you could probably do 

```
app.get_config_formatter_base()->valueSeparator(':');
```

and that would handle anything on a the top level.   
But dealing with the whitespace and array control in yaml is very different from the existing parsers so you would probably need a custom parser.  There are a few examples for JSON, so if you already had a yaml parser in your program it might be possible to modify one of the json parsers to match up with your yaml parser.  

Anyway I don't know of a clear way to handle yaml beyond trivial cases with the existing parsers or even small modifications of it.

**Reply by henryiii** [reply]
It's not just whitespace, there are a ton of things in YAML that are complex - the spec is in 10 chapters (https://yaml.org/spec/1.2.2/) and is huge. Things like tags and anchors, block and flow scalars, etc - we probably already struggle a bit with fully reading the TOML 1.0 spec, I'd hate to see what would be required for YAML! It also help that TOML tends to be used for simple configs, while YAML can be practically anything. I love YAML, but not interested in implementing it, just using it. :D

**Reply by henryiii** [reply]
I don't think SafeYAML ever took off - I _think_ it's at https://github.com/imbal/safeyaml, but maybe that's a different project. I remember the original spec that was proposed for Python included avoiding interpreting datatypes, so `3.10` would have been acceptable by the parser as a string if you requested it as a string - that would have been really nice. But TOML won over YAML or SafeYAML.

**Comment by henryiii** [comment]
We will not be able to support YAML without dependencies. That is something that should use a dependency, and would be a great "plugin" like library. But we can't support it without dependencies, as it's wildly complex. Writing it would be pretty easy, but not reading arbitrary files. I'm happy to advertise a plugin library if you want to write one, though. :)

**Reply by henryiii** [reply]
If you use a good library (haven't looked into YAML libraries for C++), writing a custom reader/writer should be really easy, and is an intentionally supported feature.

**Comment by dfleury2** [comment]
For my use cases, I am not sure I will need a full compliant YAML 1.2.2 parser, but like TOML, a single one that can parse easy config files.
(some values, and arrays not more).
Currently I am using yaml-cpp as YAML parser, but for simple things, and I am not a YAML specialist just an basic.
I will check how a reader/writer can be done in CLI11, long time since I read CLI11 code ;)

**Reply by phlptp** [reply]
I don't think the current file parser can handle any value that spans multiple lines, so it can be tweaked for the top level key value pairs and single line arrays.  anything else would take a new parser.

**Comment by dfleury2** [comment]
Not sure how to share it, but I have made an experimental Yaml config file class.
Done the bare minimum to add yam-cpp in the conanfile (requires, and generator). updating the CmakeLists.txt, addind some UT.
All is here : https://github.com/dfleury2/CLI11 (on master branch)
Only the from_config is done.

[EDIT] Added basic option for YAML in conanfile and CMakeLists.txt  (conan create failed due to test package. I do not know how to fix that)
